### PR TITLE
audit: add new homebrew/core audits

### PR DESCRIPTION
**Pull request by jonchang**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

These homebrew/core audits check for:

* `make check`
* `# tag "xxx"` comments from external taps
* `build.with?` implying the presence of options

### CODE REVIEW

File: Library/Homebrew/rubocops/lines.rb [hunk 1]
```diff
@@ -64,6 +64,15 @@ def audit_formula(_node, _class_node, _parent_class_node, _body_node)
 
             problem "Commented-out dependency #{Regexp.last_match(1)}"
           end
+
+          return unless formula_tap == "homebrew-core"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/lines.rb]:
```suggestion
          return if formula_tap != "homebrew-core"
```

File: Library/Homebrew/rubocops/lines.rb [hunk 2]
```diff
@@ -148,6 +157,16 @@ def audit_formula(_node, _class_node, _parent_class_node, body_node)
 
             problem "Reference '#{match[1]}' without dashes"
           end
+
+          return unless formula_tap == "homebrew-core"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/lines.rb]:
```suggestion
          return if formula_tap != "homebrew-core"
```

File: Library/Homebrew/rubocops/lines.rb [hunk 3]
```diff
@@ -400,6 +419,21 @@ def audit_formula(_node, _class_node, _parent_class_node, body_node)
 
             problem "Use the `#{match}` Ruby method instead of `#{method.source}`"
           end
+
+          return unless formula_tap == "homebrew-core"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/lines.rb]:
```suggestion
          return if formula_tap != "homebrew-core"
```

File: Library/Homebrew/rubocops/lines.rb [hunk 4]
```diff
@@ -64,6 +64,15 @@ def audit_formula(_node, _class_node, _parent_class_node, _body_node)
 
             problem "Commented-out dependency #{Regexp.last_match(1)}"
           end
+
+          return unless formula_tap == "homebrew-core"
+
+          audit_comments do |comment|
+            # Citation and tag comments from third-party taps
+            next unless comment =~ /#\s*(cite(?=\s*\w+:)|doi(?=\s*['"])|tag(?=\s*['"]))/
```
Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/lines.rb]:
```suggestion
            next if comment !~ /#\s*(cite(?=\s*\w+:)|doi(?=\s*['"])|tag(?=\s*['"]))/
```

File: Library/Homebrew/rubocops/lines.rb [hunk 5]
```diff
@@ -148,6 +157,16 @@ def audit_formula(_node, _class_node, _parent_class_node, body_node)
 
             problem "Reference '#{match[1]}' without dashes"
           end
+
+          return unless formula_tap == "homebrew-core"
+
+          # Use of build.with? implies options, which are forbidden in homebrew/core
+          find_instance_method_call(body_node, :build, :without?) do |_method|
```
Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/lines.rb]:
```suggestion
          find_instance_method_call(body_node, :build, :without?) do
```

File: Library/Homebrew/rubocops/lines.rb [hunk 6]
```diff
@@ -148,6 +157,16 @@ def audit_formula(_node, _class_node, _parent_class_node, body_node)
...
+            problem "Formulae in homebrew/core should not use `build.without?`."
+          end
+          find_instance_method_call(body_node, :build, :with?) do |_method|
```
Comment (MikeMcQuaid) [inline Library/Homebrew/rubocops/lines.rb]:
```suggestion
          find_instance_method_call(body_node, :build, :with?) do
```

**Comment by jonchang** [comment]
Merging since tests are 🍏 . Thanks for the review @MikeMcQuaid!

**Comment by fxcoudert** [comment]
We're seeing this in https://github.com/Homebrew/homebrew-core/pull/48770:

```
 ==> brew audit p11-kit --online
 ==> FAILED
 Error: 1 problem in 1 formula detected
 p11-kit:
   * C: 43: col 5: Formulae in homebrew/core should not run build-time checks
```

But build-time checks are allowed in security software (unless we have changed our position on that). Should that audit be only for new formulas?

**Comment by jonchang** [comment]
I didn’t know we had that policy. Two options that I can think of:

* new formulae only 
* add a list to exclude formulae from this check

I slightly prefer the latter. Thoughts?

**Comment by fxcoudert** [comment]
I'm generally against broad strokes of audit for things that we can have good reasons to ignore. And white lists need to be maintained, and new formulas added, etc. I like the "new formula" because once a maintainer decides that it's a valid use, then it's done. Simple.

If you go with a whitelist, I'd say exclude:

- Some bignum libraries (gmp mpfr), which are used in several places, regularly trigger miscompilations that we're happy to see before they hit us in unexpected places.
- Security / crypto software

**Pull request merged**